### PR TITLE
refactor(pkg/trie/triedb): introduce `HashDB` interface and integrate into `TrieDB` with added iterators

### PR DESCRIPTION
**Pull request by timwu20**
### `hashdb.HashDB`

`hashdb.HashDB` is an interface introduced around a persistent store where the hash of the inserted value is used as the key.  This interface is found prominently in substrate, and is used as the underlying storage of `triedb.TrieDB`.  The integration of this interface as the underlying storage is included with the changes.  

`hashdb.Prefix` is introduced to represent prefix partials where there could be an existence of a padded byte.  This is useful to represent nibble offset for a partial key.  

### `memorydb.MemoryDB`

`memorydb.MemoryDB` is an implementation of `HashDB` with generic type parameters for the `Hash`, `Hasher`, `Key`, and `KeyFunction`.  Variations of `memorydb.MemoryDB` using different key functions where the prefix is actually used vs not used are found in the substrate primitives and are used in PR #4318.   One key feature that is of note is that values are actually reference counted, and not actually removed unless an equal amount of removes to the amount of gets.

### `triedb.TrieDBIterator` and `triedb.TrieDBKeyIterator`

These two iterator types are introduced and used in PR #4318.  They are simply exported iterators that will iterate through the pairs (key, values) or just the keys in lexographic order.  

### `triedb.TrieDB.LookupFirstDescendant`

This method is required to for `TrieBackend` work in PR #4318.  It will look up the first descendant node from a given key.

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues
contributes to #3901

### CODE REVIEW

File: internal/hash-db/hash_db.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// Therefore it is always the leftmost portion of the node key, so its internal representation
+// is a non expanded byte slice followed by a last padded byte representation.
+// The padded byte is an optional padded value.
+type Prefix struct {
+	Key    []byte
+	Padded *byte
+}
+
+// An empty prefix constant.
+// Can be use when the prefix is not use dinternally or for root nodes.
```
Comment (haikoschol) [inline internal/hash-db/hash_db.go]:
```suggestion
// Can be use when the prefix is not used internally or for root nodes.
```

File: internal/hash-db/hash_db.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	// Insert a datum item into the DB and return the datum's hash for a later lookup. Insertions
+	// are counted and the equivalent number of remove()s must be performed before the data
+	// is considered dead.
+	Insert(prefix Prefix, value []byte) Hash
+
+	// Like Insert(), except you provide the key and the data is all moved.
+	Emplace(key Hash, prefix Prefix, value []byte)
+
+	// Remove a datum previously inserted. Insertions can be "owed" such that the same number of
+	// inserts may happen without the data being eventually being inserted into the DB.
```
Comment (haikoschol) [inline internal/hash-db/hash_db.go]:
```suggestion
	// inserts may happen without the data eventually being inserted into the DB.
```

File: pkg/trie/triedb/cache.go [hunk 3]
```diff
@@ -321,9 +352,9 @@ func (ecv ExistingCachedValue[H]) hash() *H     { return &ecv.Hash } //nolint:un
 //
 // The interface consists of two cache levels, first the trie nodes cache and then the value cache.
 // The trie nodes cache, as the name indicates, is for caching trie nodes as [CachedNode]. These
-// trie nodes are referenced by their hash. The value cache is caching [CachedValue]s and these
+// trie nodes are referenced by their hash. The value cache is caching [CachedValue]'s and these
 // are referenced by the key to look them up in the trie. As multiple different tries can have
-// different values under the same key, it is up to the cache implementation to ensure that the
+// different values under the same key, it up to the cache implementation to ensure that the
```
Comment (haikoschol) [inline pkg/trie/triedb/cache.go]: Wasn't the "is" here correct?

File: pkg/trie/triedb/triedb.go [hunk 4]
```diff
@@ -277,12 +276,14 @@ func (t *TrieDB[H, Hasher]) insert(keyNibbles nibbles.Nibbles, value []byte) err
 }
 
 // Put inserts the given key / value pair into the trie
```
Comment (haikoschol) [inline pkg/trie/triedb/triedb.go]:
```suggestion
// Set inserts the given key / value pair into the trie
```

File: pkg/trie/triedb/triedb.go [hunk 5]
```diff
@@ -358,15 +359,16 @@ type inspectResult struct {
...
 
-// inspect inspects the given node `stored` and calls the `inspector` function
+// inspect inspects the given node stored and calls the inspector function
 // then returns the new node and a boolean indicating if the node has changed
 func (t *TrieDB[H, Hasher]) inspect(
 	stored StoredNode,
 	key *nibbles.Nibbles,
 	inspector func(Node, *nibbles.Nibbles) (action, error),
 ) (*inspectResult, error) {
 	// shallow copy since key will change offset through inspector
```
Comment (haikoschol) [inline pkg/trie/triedb/triedb.go:369]:
```suggestion
```
assuming `key.Clone()` does not return a shallow copy

File: pkg/trie/triedb/triedb.go [hunk 6]
```diff
@@ -358,15 +359,16 @@ type inspectResult struct {
...
-	currentKey := *key
+	// currentKey := *key
```
Comment (haikoschol) [inline pkg/trie/triedb/triedb.go]:
```suggestion
```

**Review by haikoschol — APPROVED** [review]

File: pkg/trie/triedb/node.go [hunk 7]
```diff
@@ -108,9 +109,9 @@ func (vr newValueRef[H]) equal(other nodeValue) bool {
 }
 
 func NewValue[H hash.Hash](data []byte, threshold int) nodeValue {
-	if len(data) >= threshold {
+	if len(data) > threshold {
```
Comment (dimartiro) [inline pkg/trie/triedb/node.go:112]: good catch, do we have a test for this?

**Review by dimartiro — APPROVED** [review]
LGTM

File: pkg/trie/triedb/triedb.go (hunk 5 above)
Comment (timwu20) [reply pkg/trie/triedb/triedb.go:369]: from the `slices.Clone` functionality, they use assignment so I would consider it a shallow clone.  The actual offset is copied via assignment as well so I would consider that as a shallow copy.

File: pkg/trie/triedb/node.go (hunk 7 above)
Comment (timwu20) [reply pkg/trie/triedb/node.go:112]: Yea this is called repeatedly via `TrieDB` methods.  I actually uncovered this problem through the tests.

**Pull request merged**